crates/circuit/src/parameter/symbol_expr.rs

```diff
@@ -622,26 +671,39 @@ impl SymbolExpr {
     }
 
     /// Return hashset of all parameters this equation contains.
-    pub fn parameters(&self) -> HashSet<Symbol> {
+    /// TODO This should just be iter_symbols.collect()
+    pub fn symbols(&self) -> HashSet<Symbol> {
         match self {
             SymbolExpr::Symbol(e) => HashSet::from_iter([e.clone()]),
             SymbolExpr::Value(_) => HashSet::<Symbol>::new(),
-            SymbolExpr::Unary { op: _, expr } => expr.parameters(),
+            SymbolExpr::Unary { op: _, expr } => expr.symbols(),
             SymbolExpr::Binary { op: _, lhs, rhs } => {
                 let mut parameters = HashSet::<Symbol>::new();
-                for s in lhs.parameters().union(&rhs.parameters()) {
+                for s in lhs.symbols().union(&rhs.symbols()) {
                     parameters.insert(s.clone());
                 }
                 parameters
             }
         }
     }
 
+    /// This could maybe be more elegantly resolved with a SymbolIter type>
+    pub fn iter_symbols(&self) -> Box<dyn Iterator<Item = Symbol>> {
+        match self {
+            SymbolExpr::Symbol(e) => Box::new(::std::iter::once(e.clone())),
+            SymbolExpr::Value(_) => Box::new(::std::iter::empty()),
+            SymbolExpr::Unary { op: _, expr } => expr.iter_symbols(),
+            SymbolExpr::Binary { op: _, lhs, rhs } => {
+                Box::new(lhs.iter_symbols().chain(rhs.iter_symbols()))
+            }
+        }
+    }
+
     /// Map of parameter name to the parameter.
     pub fn name_map(&self) -> HashMap<String, Symbol> {
-        self.parameters()
+        self.symbols()
```

Comment:
This is a perfect example of what I was talking about before with the hidden overhead with these methods. This allocates twice into a `HashMap`, `symbols()` collects into a `HashSet` and then we in turn just iterate over that into the output hashmap. You should be able to remove one of these allocations.

Reply:
done by replacing this with `iter_symbols` directly 👍🏻